### PR TITLE
Library concurency refactoring

### PR DESCRIPTION
**Pull request by MK-42**
mixxxdj/mixxx#73

Trojans GSOC project (Async DB access) reworked to split TrackCollection to public and private part.

### CODE REVIEW

**Comment by MK-42** [comment]
This won't compile because of the Library-Scanner. It is not reworked now, as I don't understand It yet.

It uses its own DAOs - why? And sometimes it uses those from TrackCollection - Why?

**Comment by MK-42** [comment]
@troyane would you mind to look at my questions about your LibraryScanner changes? Some are not clear to me.

**Comment by daschuer** [comment]
> It uses its own DAOs - why?

Because it uses currently a different thread. This way sq-light allows concurrent access. 
But it is locking, which has introduced some deadlocks in the past, hopfully solved in master.

**Comment by daschuer** [comment]
My be it is a good idea to remove the extra instance for the library scanner and use always the lambda. 
This will make the code clearer and should not suffer performance issues since the database access is locking anyway. 
We need to verify this though.

**Comment by MK-42** [comment]
I'm currently facing some nested callSync calls which cause a deadlock. Need to investigate this to get it running again

**Comment by daschuer** [comment]
This seams to be also a common issue that may happens later as well. 
Can we place flag and an assertion that guards against this condition?

**Comment by MK-42** [comment]
>Can we place flag and an assertion that guards against this condition?

There is one.

**Comment by MK-42** [comment]
well those nested calls get tricky... that will take a while

**Comment by daschuer** [comment]
Can you give an example?

**Comment by MK-42** [comment]
Maybe look at 3add7af6ac164ad0c9944432add66d328a83969f there is one of them fixed
I'm also facing some problems with slots containing a lambda and beeing called from one. I thought Qt::QueuedConnection would help there but it doesn't

**Comment by MK-42** [comment]
Not its segfaulting. What a crappy little thing

**Comment by MK-42** [comment]
Well in fact it is a infinite loop in lookupCachedTrack, happening as soon as a track is in the library. screw...

**Comment by daschuer** [comment]
Do you have the deadlocks in the original files as well? 
I have tried to track the issue from 3add7af, but I failed. 
Qt should choose queued connections when it jumps across thread borders. 
... Or does the connect call inside a Lambda confuse Qt? 

Are you able to single step into the Qt code? If not it might be worth to set it up.  

An other thought: 
Be aware that the slots from the DAOs are also called directly from the GUI thread. Since they inherit the signal message queue from the parent Object. 
If we need slots that are executed in the Track Collection thread, we may execute them from an new 
signal message queue.  
See: moveToThread() and QCoreApplication::processEvents()

**Comment by daschuer** [comment]
Does any of your commit build? I would like to test. 
What is the receipt to get the deadlock?

**Comment by MK-42** [comment]
It builds. Just Start it. You will geht some Kind Of race condition to a segfault at the Moment i think if im not wrong. Cant look into it right now

Am 17. Juni 2015 22:47:21 MESZ, schrieb "Daniel Schürmann" <notifications@github.com>:
>Does any of your commit build? I would like to test. 
>What is the receipt to get the deadlock? 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mixxxdj/mixxx/pull/629#issuecomment-112945698

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by daschuer** [comment]
Building fails on my site:  

```
src/test/searchqueryparsertest.cpp: In constructor ‘SearchQueryParserTest::SearchQueryParserTest()’:
src/test/searchqueryparsertest.cpp:12:34: error: no matching function for call to ‘SearchQueryParser::SearchQueryParser(QSqlDatabase&)’
               m_parser(m_database) {
                                  ^
src/test/searchqueryparsertest.cpp:12:34: note: candidates are:
In file included from src/library/searchqueryparser.h:8:0,
                 from src/test/searchqueryparsertest.cpp:6:
src/library/searchqueryparser.h:54:30: note: SearchQueryParser::SearchQueryParser(const SearchQueryParser&)
     DISALLOW_COPY_AND_ASSIGN(SearchQueryParser);
                              ^
src/util.h:5:3: note: in definition of macro ‘DISALLOW_COPY_AND_ASSIGN’
   TypeName(const TypeName&);               \
   ^
src/library/searchqueryparser.h:54:30: note:   no known conversion for argument 1 from ‘QSqlDatabase’ to ‘const SearchQueryParser&’
     DISALLOW_COPY_AND_ASSIGN(SearchQueryParser);
                              ^
src/util.h:5:3: note: in definition of macro ‘DISALLOW_COPY_AND_ASSIGN’
   TypeName(const TypeName&);               \
   ^
In file included from src/test/searchqueryparsertest.cpp:6:0:
src/library/searchqueryparser.h:13:5: note: SearchQueryParser::SearchQueryParser(TrackCollection*)
     SearchQueryParser(TrackCollection* pTrackCollection);
     ^
src/library/searchqueryparser.h:13:5: note:   no known conversion for argument 1 from ‘QSqlDatabase’ to ‘TrackCollection*’
scons: *** [lin64_build/test/searchqueryparsertest.o] Error 1
scons: building terminated because of errors.
```

**Comment by MK-42** [comment]
Sorry, i build without the tests

Am 17. Juni 2015 23:09:28 MESZ, schrieb "Daniel Schürmann" <notifications@github.com>:
>Building fails on my site:  
>
>```
>src/test/searchqueryparsertest.cpp: In constructor
>‘SearchQueryParserTest::SearchQueryParserTest()’:
>src/test/searchqueryparsertest.cpp:12:34: error: no matching function
>for call to ‘SearchQueryParser::SearchQueryParser(QSqlDatabase&)’
>               m_parser(m_database) {
>                                  ^
>src/test/searchqueryparsertest.cpp:12:34: note: candidates are:
>In file included from src/library/searchqueryparser.h:8:0,
>                 from src/test/searchqueryparsertest.cpp:6:
>src/library/searchqueryparser.h:54:30: note:
>SearchQueryParser::SearchQueryParser(const SearchQueryParser&)
>     DISALLOW_COPY_AND_ASSIGN(SearchQueryParser);
>                              ^
>src/util.h:5:3: note: in definition of macro ‘DISALLOW_COPY_AND_ASSIGN’
>   TypeName(const TypeName&);               \
>   ^
>src/library/searchqueryparser.h:54:30: note:   no known conversion for
>argument 1 from ‘QSqlDatabase’ to ‘const SearchQueryParser&’
>     DISALLOW_COPY_AND_ASSIGN(SearchQueryParser);
>                              ^
>src/util.h:5:3: note: in definition of macro ‘DISALLOW_COPY_AND_ASSIGN’
>   TypeName(const TypeName&);               \
>   ^
>In file included from src/test/searchqueryparsertest.cpp:6:0:
>src/library/searchqueryparser.h:13:5: note:
>SearchQueryParser::SearchQueryParser(TrackCollection*)
>     SearchQueryParser(TrackCollection* pTrackCollection);
>     ^
>src/library/searchqueryparser.h:13:5: note:   no known conversion for
>argument 1 from ‘QSqlDatabase’ to ‘TrackCollection*’
>scons: *** [lin64_build/test/searchqueryparsertest.o] Error 1
>scons: building terminated because of errors.
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mixxxdj/mixxx/pull/629#issuecomment-112950351

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by daschuer** [comment]
Oh, my bad. Will try without the tests.

**Comment by daschuer** [comment]
This branch already includes a single DB access point for GUI and LibraryScanner thread. 
This was introduces in 8c6dcd0dab1d89a0ed2d120adef1a40f099707d2
This means you need not distinguish between Library scanner and GUI. 

Currently The library scanner hold its own instances of these 
    m_libraryHashDao.setDatabase(m_database);
    m_cueDao.setDatabase(m_database);
    m_trackDao.setDatabase(m_database);
    m_playlistDao.setDatabase(m_database);
    m_analysisDao.setDatabase(m_database);
I think we can remove them after having a close look if there member variables interfere.
Race conditions cant be the issue, because all access is done in a single thread once we are finished. 

I think this way we get around the dangerous m_database member. 
 
The current commit suffers a nesting callSync. Are you just fixing it or do you need help?

**Comment by MK-42** [comment]
>The current commit suffers a nesting callSync. Are you just fixing it or do you need help?

I tried to fix it but currently with no success. My general plan is to pass over the TrackCollectionPrivate to where it is needed in the callstack. The nested calls are most often because of me having added a callSync to get the TCPrivate in that context.

**Comment by MK-42** [comment]
To make it clear: Yes - any help is appreciated, I'm short on time this weekend

**Comment by MK-42** [comment]
We have a warning at startup:
```
Debug [Main]: callSync from "virtual void SetlogFeature::buildPlaylistList()" 
Warning [TrackCollection]: QObject: Cannot create children for a parent that is in a different thread.
(Parent is SetlogFeature(0x1df4f70), parent's thread is QThread(0x123a920), current thread is TrackCollection(0x1dad2d0)
```

that needs investigation

File: src/musicbrainz/gzip.cpp [hunk 1]
```diff
@@ -40,7 +40,7 @@ QByteArray gzipCompress(const QByteArray &data)
 	compressedData.remove(compressedData.size() - 4, 4); // zlib footer
 
 	QByteArray result;
-	result.append(header, 10);
+	result.append(reinterpret_cast<const char*>(header), 10);
```
Comment (daschuer) [inline src/musicbrainz/gzip.cpp]: reinterprete_cast looks evil, use static_cast in such a case. 
However we have append(const char * str) so the original version was not that bad. 
What was the error message? Probably that the array initializer are int.  

One other problem is that char can either be signed or unsigned. It is compiler specific. 

http://en.cppreference.com/w/cpp/language/character_literal

According to this a cast free solution will look like: 

```
	const unsigned char header[10] = {
		'\x1f', '\x8b', // ID1 + ID2
		'\8', // Compression Method
		'\0', // Flags
		'\0', '\0', '\0', '\0', // Modification Time
		'\2', // Extra Flags
		'\255', // Operating System
	};
```
Can you try this?

Comment (MK-42) [inline src/musicbrainz/gzip.cpp]: This code is Copy paste from master Branch.
The error was an Wnarrowing that 255 and another number are int.

Any change to that should go to master, not here i think

Am 7. Juli 2015 08:27:32 MESZ, schrieb "Daniel Schürmann" <notifications@github.com>:
>> @@ -40,7 +40,7 @@ QByteArray gzipCompress(const QByteArray &data)
>>  	compressedData.remove(compressedData.size() - 4, 4); // zlib footer
>>  
>>  	QByteArray result;
>> -	result.append(header, 10);
>> +	result.append(reinterpret_cast<const char*>(header), 10);
>
>reinterprete_cast looks evil, use static_cast in such a case. 
>However we have append(const char * str) so the original version was
>not that bad. 
>What was the error message? Probably that the array initializer are
>int.  
>
>One other problem is that char can either be signed or unsigned. It is
>compiler specific. 
>
>http://en.cppreference.com/w/cpp/language/character_literal
>
>According to this a cast free solution will look like: 
>
>```
>	const unsigned char header[10] = {
>		'\x1f', '\x8b', // ID1 + ID2
>		'\8', // Compression Method
>		'\0', // Flags
>		'\0', '\0', '\0', '\0', // Modification Time
>		'\2', // Extra Flags
>		'\255', // Operating System
>	};
>```
>Can you try this? 
> 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mixxxdj/mixxx/pull/629/files#r34010823

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

Comment (daschuer) [inline src/musicbrainz/gzip.cpp]: Yes, this will avoid conflicts later.

**Comment by MK-42** [comment]
I've tracked down the Caching problem (Logfile-Spamming) by introducing critical changes in tro's state of the branch.

Its src/library/basetrackcache.cpp:155. When introducing a lambda in tro's code, spamming happens also in his branch. IMHO we need a lambda here because of access to DAO (in the current state there is also no way around it, because we need the TCPrivate to access it).

As you know I'm totally new to mixxxxs codebase and have no clue why moving that statement syncronously into another thread causes that spamming.

**Comment by MK-42** [comment]
Pulling the if condition on m_bIsCaching into the lambda doesn't solve it.

**Comment by daschuer** [comment]
Mixxx has two kind of track caches. A library column cache BaseTrackCache::m_trackInfo and a track object cache used everywhere in Mixxx TrackDAO::m_trackCache. The later one allways contains the most recent values of metadata. So BaseTrackCache looks there first, before using its own cached values.  

If I skip the first lookup of the TrackDAO::m_trackCache, the logfile spamming stops. In this solution, the library shows allays the persistant DB values of the track and not the values just modified and not stored. 

I cannot sea a reason why you lamda solution starts this endless lookups, but in any case we must keep the TrackDAO::m_trackCache fast. So I think we need to move it out of the TrackDAO and keep it accessible from any thread. 

Normally the Library view lookups are generated by any Mouse events, but should stop if Mixxx is idle. 
We should investigate why we cant make it stop with the lamda in place. 
Unfortunately a breakpoints tends to triggers the lookup :-/

**Comment by daschuer** [comment]
Got it! It is TrackCollection::setUiEnabled() which triggers the repaint.

**Comment by MK-42** [comment]
Well done!
So what to do now?

**Comment by daschuer** [comment]
I am currently working on the next merge: Banshee feature. 

From the merge workflow, it would be the best to postpone big refactorings to the time we are again head up with master. This would mean: just comment out TrackCollection::setUiEnabled() and continue to test and fix the remaining small issues. 
I remember we had some changes in the way caching works, which are probably not merged.

**Comment by MK-42** [comment]
May I ask how your workflow is for those merges?
I would skim through the git log, searching for the last merged commit, search the next commit with something like "merge PR #xy from.." and do a git merge $(Hash from that commit).
Is this how you are working or is there an easier way?

**Comment by daschuer** [comment]
Yes, that is how I work. I use these tools: 
* gitk master LibraryConcurencyRefactoring -> for a nice view of both branches 
* meld -> for viewing diffs 
* smartgit -> for editing conflicts using meld as diff viewer 

I think I will finish the Banshee merge tonight. Would you like to adopt the merge job?

**Comment by MK-42** [comment]
I would do some of the work, sure.
I will be busy the next 6 weeks because of exams and some holiday work, so I can't start just now but will if I have more time after that. Of course I will be happy if you have time to do some merges until then.

**Comment by MK-42** [comment]
in `src/analyserwaveforms.cpp` is another db_connection. Is this a problem with QSql because its in a diffrent thread then TrackCollection?

**Comment by MK-42** [comment]
@daschuer would you mind reviewing 7b6dd6c about executing lambdas directly in the TrackCollection Thread?
I've also done some merges. The next is multiple LibraryFolders. I won't have time for that till I come back in some weeks.

**Comment by daschuer** [comment]
Works like a charm! Thank you.

**Comment by MK-42** [comment]
I forgot to mention: If you want to take over the merging job till I come back - you are of course welcome to do so.

**Comment by MK-42** [comment]
Welcome in Year 2014 again :) still 1 year to go. I'ts compiling the next one right now. The following is a bansee_feature branch. I'm scared ;-) But works well so far

**Comment by MK-42** [comment]
Enough for today. BansheeFeature is really heavy. will need much time

**Comment by daschuer** [comment]
Nice progress!
Wasn't 49d3be619f68b3b5f3d580411ea8e30fa415cb94 the banshee branch? 
I have a almost working branch up to 6c15845612d4b3d1222b6dee12890b0aab3a9e36
It is #21 but I think you have passed me. 
I hope I find time to look at this soon.

**Comment by MK-42** [comment]
There is another merge of your feature branch banshee_feature (now the next one in the history) which conflicts a lot. This will take a while to merge.

IF you have any bugfixes, feel free to post a PR or do the next merges.
Let's anounce here if we are starting to merge again to prevent additional work for both of us.

**Comment by MK-42** [comment]
Now we have ~400PRs to go...

**Comment by daschuer** [comment]
Yeehaw :-)

**Comment by kain88-de** [comment]
https://github.com/mhagger/git-imerge

This tool might help you with the merge

**Comment by MK-42** [comment]
I just extended it to call scons after every auto merge. Looks handy.
At the moment I'm stuck at a commit thats reworking the library scan for another round... Merging the library-scanning code is very hard but I'll do my best.

**Comment by MK-42** [comment]
So I've done some more merges which are compiling fine. The Library scanner is broken at the moment, it only scans the first subfolder of the music library (at least on my machine).

@daschuer  and others:
The LibraryScanner Thread closed the database what crashed nearly everything inside mixxxx afterwards. Should the scanner operate on its own instance of m_database or the one which is also used by TrackCollectionPrivate? As  don't know much about the library scanner I don't know at the moment.

**Comment by daschuer** [comment]
I hope I can recall it right: 

We had deadlock issues in the past because the library scanner is on its own thread and uses its own database instance. So I think it is better to use only one database instance and one database thread. 

By the way: Thank you very much for being back :-)

**Comment by MK-42** [comment]
There has been so much work done in the LibraryScanner in both threads that I can't figure out completely how it should work correctly. It is at the moment not working correctly - it is only adding the files of the first subdirectory but hashing all directories. So something in the recursion is probably wrong. Someone with deeper kowledge has to look into that. But maybe we should rewrite/reinvestigate the whole thing when the merging is done.
What do you think @daschuer?

**Comment by daschuer** [comment]
I hope I will find time to look at it soon. 

There will be for sure still a lot of work, once the merge is done. So it would be OK for me, to finish the merge, without much testing, just keep the beast compiling. 
Once it is head up with master, we can verify all touched files by a diff against master and verify if all touched features are still working .. or make it work again.

**Comment by MK-42** [comment]
Well... some work is done. about 250PRs from now on and we are back at master.
Actually we are back in 31 October 2014 now, so its getting faster atm.

**Comment by daschuer** [comment]
Good job. Thank you. I hope you do not suffer some kind of git madness after you are finished ;-). 

Small diffs are always a good idea!
This is already 4k+ :-/ but hopefully we are able to reduce it once we are back at master. 

> Libraryscanner:

You ideas seams to be reasonable. 

I just wonder what will be the right approach to get this one merged, without breaking too much. 

Here some ideas: 

This branch is somehow the truth so we we should keep it in any case for reference. Because it contains the real git history with all merge commits in place, it is unfortunately a bit  unhandy to work with. For example, you cannot rebase it without doing all merges again.

To solve this we should set up a second branch which, still contains all the commit, but "git filtered" to drop the second parent, turning all merges into normal commits. From this new branch we are able to do all the git magic that might be required to split squash, reorder or whatever to build smaller pull requests, that are merge-able and review-able.  

I am curious to see how we will finally continue.

**Comment by MK-42** [comment]
I Thought about reparenting the last merge commit to my first merging commit. The information in between is not really worth the mess. When we review that thing we will probably look only at the diff against master, not the intermediate commits.
And I don't think that there are other branches that are that old which would lead to merge this beast again with?

So I think a second branch which contains the info of this one and the last commit reparented to the first merge could do the trick.

**Comment by daschuer** [comment]
> The information in between is not really worth the mess.

We will see, how hard it is to fix all bugs. Have this branch in place can't hurt.
Probably we are able to find a solution for a hard problem using it. 

Squashing all merge commits to one, in a second branch is a good idea. Finally no one needs these hug number of merge commits in the master branch.

**Comment by MK-42** [comment]
Looking back at this.. wow. From my current perspective, I think its not feasible to go that way with merging further without loosing the head in there. IIRC, the library scanner and other stuff is already heavily broken by that merging here, as one is not able to test that stuff after each merge.
Thinking about how that situation here could be resolved I will collect my thoughts, looking for feedback:

IIRC we have to make sure that only a single thread accesses QSql related stuff, and allow to put some more work in asyncronous ways out of the GUI-thread.

Maybe we should think about a more step-by-step approach. This PR is aiming to push the db-access to a single thread and also introducing async db access. Maybe we should do it step by step, first consolidate the db-access in a single entry point somewhere in the code, and after that is proven to work implement some asynchronous access methods where needed. I think that will also make the diff smaller and easier to review.

Looking at @uklotzde's work on the crate database access in #1099, it could also be a possible solution, to handle the threading problems inside of the FwdSqlQuery or subclasses of it. We could move the execution of that query into a dedicated Database-thread, introduce proper locking there, and get the rest of the code use such derived QSqlQuery classes. Probably not that much has to change, as we can keep the api stable for the query (as long as we stay with syncronus access). We only need to make sure that the code doesn't use plain QSqlQueries from that point on. That point could be merged as is, making smaller steps into master, keeping the diff small.

After that is done, we could introduce a Async(Fwd)SqlQuery, that tells it's caller when it is done (probably via callback), and allow him to fetch the data then. I don't know if the real bottleneck for the current responsibility issues with mass-actions is the sql-query or the postprocessing of it. If its postprocessing, we could also allow other code than sql-queries (lambdas?) to be queued into the library-thread, to allow sorting, building tables and stuff from non-gui-thread. For that, trojans queuing and locking code could be reused.

Btw.: as 2.1 is on i'ts way I don't want to get anything in there, and don't want to bind dev-ressources anywhere non-necessary for 2.1. But starting to think on that and get to a conclusion how to do it in small, master-mergeble steps would be good I think.

**Comment by uklotzde** [comment]
I don't think that there is a chance to merge this PR. And I'm not convinced that the approach will work as expected.

**Comment by daschuer** [comment]
Yes, this one was rotting for too long, I am really sorry. 
Do you mean that the lambda approach will not work? We had good results during the GSoC project. 
And on top of the latest refactorings, it should work even better. 
What could be an alternative solution?

**Comment by uklotzde** [comment]
Even with lambdas you still need to care about thread-safety and concurrency issues. Wrapping arbitrary code with references to GUI classes into lambdas and executing them in another thread's context sometimes later does not sound like a good idea to me. It may be fast, but is is correct and safe?

Qt's queued signal/slot connections might be useful here. Passing immutable data from one thread is at least safe, albeit with some overhead. I already started to separate the main track select query with such an approach, but didn't have time to complete it, yet. Stay tuned ;)

**Comment by Be-ing** [comment]
This pull request has not been updated in years, has accumulated tons of merge conflicts, and there seem to be concerns about its fundamental approach so I am closing it to reduce the clutter of open PRs.

**Pull request closed without merging**